### PR TITLE
MultivariatePolynomial

### PR DESCRIPTION
**Pull request by StoicBronco**
This contains just the MultivariatePolynomial class on its own; this class is also contained in PR #913.
@isuruf @shivamvats @chenchfort @certik @Sumith1896 @irislq

### CODE REVIEW

**Comment by myluszczak** [comment]
Currently, this build fails only Travis CI tests 13 and 14.  I restarted the builds on my own repository and they passed.  Could someone restart the builds for this PR?

File: symengine/dict.h [hunk 1]
```diff
@@ -174,6 +179,122 @@ int map_int_Expr_compare(const map_int_Expr &a, const map_int_Expr &b);
...
+int umap_uvec_mpz_compare(const umap_uvec_mpz &a, const umap_uvec_mpz &b);
+
+// copied from umap_eq, with derefrencing of image in map removed.
+bool umap_uvec_mpz_eq(const umap_uvec_mpz &a, const umap_uvec_mpz &b);
+
+int umap_uvec_expr_compare(const umap_uvec_expr &a, const umap_uvec_expr &b);
+bool umap_uvec_expr_eq(const umap_uvec_expr &a, const umap_uvec_expr &b);
+
+template <class T>
+bool set_eq(const T &A, const T &B)
```
Comment (isuruf) [inline symengine/dict.h]: Use `vec_set_eq`

File: symengine/dict.h [hunk 2]
```diff
@@ -174,6 +179,122 @@ int map_int_Expr_compare(const map_int_Expr &a, const map_int_Expr &b);
...
+    auto b = B.begin();
+    for (; a != A.end(); ++a, ++b) {
+        if (neq(**a, **b))
+            return false; // values not equal
+    }
+    return true;
+}
+
+template <class T>
+int set_compare(const T &A, const T &B)
```
Comment (isuruf) [inline symengine/dict.h]: Use `vec_set_compare`

File: symengine/dict.h [hunk 3]
```diff
@@ -174,6 +179,122 @@ int map_int_Expr_compare(const map_int_Expr &a, const map_int_Expr &b);
...
+    {
+        std::size_t h = 0;
+        for (unsigned int i : v) {
+            h ^= i + 0x9e3779b + (h << 6) + (h >> 2);
+        }
+        return h;
+    }
+};
+
+class vec_uint_eq
```
Comment (isuruf) [inline symengine/dict.h]: This is unnecessary

Comment (isuruf) [inline symengine/dict.h]: Use `==` of `std::vector`

File: symengine/dict.h [hunk 4]
```diff
@@ -174,6 +179,122 @@ int map_int_Expr_compare(const map_int_Expr &a, const map_int_Expr &b);
...
+            return false;
+        for (unsigned int i = 0; i < a.size(); i++) {
+            if (a[i] != b[i])
+                return false;
+        }
+        return true;
+    }
+};
+
+class vec_uint_compare
```
Comment (isuruf) [inline symengine/dict.h]: Use `<` of `std::vector`

**Comment by isuruf** [comment]
For all the functions, you have special cases for all univariate types.
Isn't it better for univariate types to be converted multivariate types before actually doing any operation?
For example `MultivariatePolynomial` has special cases for addition with `UnivariatePolynomial`, but in the future, we need to add special cases for `UnivariateIntPolynomial` and also `MultivariateIntPolynomial`.

So, I think converting `UnivariatePolynomial` to `MultivariatePolynomial` first is desired. This also increases the maintainability of the code.

What do you think?

File: symengine/polynomial_multivariate.h [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+class MultivariateIntPolynomial : public Basic
+{
+public:
+    // vars: set of variables for the polynomial
+    // degrees: max degrees of the symbols
+    // dict: dictionary for sparse represntation of polynomial, x**1 * y**2 + 3
+    // * x**4 * y ** 5
+    // is represented as {(1,2):1,(4,5):3}
+    set_sym vars_;
+    umap_sym_uint degrees_;
```
Comment (isuruf) [inline symengine/polynomial_multivariate.h]: Where is this used?

File: symengine/symbol.h [hunk 6]
```diff
@@ -52,6 +52,33 @@ inline RCP<const Symbol> symbol(const std::string &name)
...
+};
+
+struct RCPSymbolEq {
+public:
+    bool operator()(const RCP<const Symbol> &a,
+                    const RCP<const Symbol> &b) const
+    {
+        return a->__eq__(*b);
+    }
+};
```
Comment (isuruf) [inline symengine/symbol.h]: These structs are not used right?

**Comment by myluszczak** [comment]
@isuruf There are a couple of things I'd like to discuss before merging #925. Firstly, I'd like MultivariatePolynomial to be able to have negative exponents; this should be easy to do and this change could be made either before merging or in a new PR after merging. More importantly, we were planning to move the implementation of MultivariatePolynomial to MultivariateExprPolynomial in order to make MultivariateExprPolynomial mutable; this implementation seems to make that significantly trickier.

**Comment by isuruf** [comment]
> There are a couple of things I'd like to discuss before merging #925. Firstly, I'd like MultivariatePolynomial to be able to have negative exponents; this should be easy to do and this change could be made either before merging or in a new PR after merging.

+1

> More importantly, we were planning to move the implementation of MultivariatePolynomial to MultivariateExprPolynomial in order to make MultivariateExprPolynomial mutable; this implementation seems to make that significantly trickier.

Actually, I think this implementation should make it easier. Let me know if you need any details on how to do it.

**Comment by myluszczak** [comment]
@isuruf Working off what Iris did with `UnivariateExprPolynomial`, we could use `MultivariateExprPolynomial` in place of Dict.   We'd have to give it an `insert` function and the ability to be iterated through in the loops, however.  It'd also have to have its own copy of `vars_`, since that information is necessary for preforming the arithmetic operations.  That'd somewhat torpedo the benefit of having implementations in exactly one place, however, since `MultivariateExprPolynomial` would need its own implementations of `operator+` etc.  We could rewrite `add(Mpoly)` etc to work with the overloaded operators, but that'd require writing an equivalent class for MultivariateIntPolynomial.

I guess the reason I don't see how this is supposed to work is that I don't see how to divorce the implementation of `MultivariatePolynomial` from that of `MultivariateIntPolynomial`, and I'm assuming that's necessary for what we want to do.

**Comment by isuruf** [comment]
You can't separate the two classes like in the univariate case. My suggestion is to,
1. Change the parent of `MPolyBase` from Basic to nothing
2. Rename `MultivariatePolynomial` to `MultivariatePolynomialExpr`
3. Rename `MultivariateIntPolynomial` to `MultivariateIntPolynomialExpr`
4. Add inplace operator overloads `+=`, `-=`, and so on.

Wrapper class for `MultivariatePolynomialExpr` which extends `Basic` can be figured out later after @srajangarg has finished the univariate case

**Comment by myluszczak** [comment]
@isuruf I see, thank you.

File: symengine/polynomial_multivariate.h [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+            }
+        }
+        return MPoly::from_dict(s, std::move(dict));
+    }
+};
+
+// MultivariateIntPolynomial
+class MultivariateIntPolynomial
+    : public MPolyBase<MultivariateIntPolynomial, umap_uvec_mpz, integer_class,
+                       vec_uint>
```
Comment (isuruf) [inline symengine/polynomial_multivariate.h]: Can't this also be `vec_int`?

**Comment by myluszczak** [comment]
@isuruf Yes, it could.  However, we really only need the negative exponents for `MultivariatePolynomial`.   I think that having the `Vec` template parameter would still be useful since it would allow the extension of  `MPolyBase` into polynomials with any coefficient type and any exponent type, which would be useful in implementing, say, Puiseux series.

**Comment by myluszczak** [comment]
@isuruf Looking at codecov, it looks like the main issue is that we're not using string printing in the tests anymore.  Since the printing is inconsistent across platforms, I'm not sure we can test it.  We won't need that functionality for `MultivariatePolynomialExpr`, I could just remove that if you think that'd be necessary before merging.

File: symengine/tests/printing/test_printing.cpp [hunk 8]
```diff
@@ -306,38 +305,48 @@ TEST_CASE("test_univariate_polynomial(): printing", "[printing]")
     Expression c(symbol("c"));
     Expression d(symbol("d"));
 
-    p = univariate_polynomial(x, {{0, Expression(0)}});
+    p = univariate_polynomial(x,
+                              UnivariateExprPolynomial({{0, Expression(0)}}));
```
Comment (isuruf) [inline symengine/tests/printing/test_printing.cpp]: Why this change?

File: symengine/expression.cpp [hunk 9]
```diff
@@ -9,7 +9,7 @@ namespace detail
 std::string poly_print(const Expression &x)
 {
     Precedence prec;
-    if (prec.getPrecedence(x.get_basic()) == PrecedenceEnum::Add) {
+    if (prec.getPrecedence(x.get_basic()) != PrecedenceEnum::Atom) {
```
Comment (isuruf) [inline symengine/expression.cpp]: Can you revert this change? This was added by mistake

**Comment by myluszczak** [comment]
@isuruf I think that the change you pointed out was made by reverting to a version of `test_printing.cpp` in this branch made before the merge.  I've reverted back to the version in master.

**Comment by isuruf** [comment]
Closed in favour of #950
